domain/modelmanager/service/service.go

```diff
@@ -0,0 +1,676 @@
+	ctx context.Context,
+) (string, string, error) {
+	ctx, span := trace.Start(ctx, trace.NameFromFunc())
+	defer span.End()
+
+	ctrlUUID, err := s.st.GetControllerModelUUID(ctx)
+	if errors.Is(err, modelerrors.NotFound) {
+		return "", "", errors.New(
+			"controller model does not exist to determine the default cloud and region for the controller",
+		).Add(modelerrors.NotFound)
+	} else if err != nil {
+		return "", "", errors.Errorf(
+			"getting controller model uuid: %w", err,
+		)
+	}
+
+	cloudName, regionName, err := s.st.GetModelCloudNameAndRegion(ctx, ctrlUUID)
+	if errors.Is(err, modelerrors.NotFound) {
+		return "", "", errors.New(
+			"controller model does not exist to determine the default cloud and region for the controller",
+		).Add(modelerrors.NotFound)
+	} else if err != nil {
+		return "", "", errors.Errorf(
+			"getting controller model %q to determin default cloud and region: %w",
+			ctrlUUID, err,
+		)
+	}
+	return cloudName, regionName, nil
+}
+
+// GetModelUUIDForNameAndOwner returns the model uuid for the model that exists
+// for name and owned by the supplied user.
+//
+// The following errors can be expected:
+// - [github.com/juju/juju/domain/access/errors.UserNotFound] when no user
+// exists for the supplied user name.
+// - [coreerrors.NotValid] when the user name supplied is not valid.
+// - [modelerrors.NotFound] when no model exists for the supplied name and
+// - [modelmanagererrors.ModelNameNotValid] when the model name is not valid.
+// owner.
+func (s *Service) GetModelUUIDForNameAndOwner(
+	ctx context.Context,
+	modelName string,
+	ownerName coreuser.Name,
+) (coremodel.UUID, error) {
+	ctx, span := trace.Start(ctx, trace.NameFromFunc())
+	defer span.End()
+
+	if ownerName.IsZero() {
+		return "", errors.New("owner name is required").Add(coreerrors.NotValid)
+	}
+
+	if !coremodel.IsValidModelName(modelName) {
+		return "", errors.New("model name is not valid").Add(
+			modelmanagererrors.ModelNameNotValid,
+		)
+	}
+
+	return s.st.GetModelUUIDForNameAndOwner(ctx, modelName, ownerName)
+}
+
+// getWatchActivatedModelsMapper provides a mapper for filtering out model uuid
+// changes to just those models that are activated in the controller. This func
+// will maintain the order changes received in.
+func getWatchActivatedModelsMapper(st WatchableState) eventsource.Mapper {
+	return func(
+		ctx context.Context,
+		changes []changestream.ChangeEvent,
+	) ([]changestream.ChangeEvent, error) {
+		modelUUIDs := make([]coremodel.UUID, 0, len(changes))
+		for _, change := range changes {
+			modelUUIDs = append(modelUUIDs, coremodel.UUID(change.Changed()))
+		}
+
+		filteredUUIDChanges, err := st.IdentifyActiveModelsFromList(ctx, modelUUIDs)
+		if err != nil {
+			return nil, errors.Errorf(
+				"identifying active models from change set: %w", err,
+			)
+		}
+
+		// Create a map of filtered UUIDs so we can look them up quickly.
+		filteredUUIDChangesMap := make(map[string]struct{}, len(filteredUUIDChanges))
```

Comment:
Why not use a strings Set

Reply:
It is a worse solution, it requires more memory and also is slower because it doesn't pre-allocate.